lib/dry/result_matcher/evaluator.rb

```diff
@@ -0,0 +1,40 @@
+module Dry
+  module ResultMatcher
+    class Evaluator
+      attr_reader :result
+      attr_reader :output
+
+      def initialize(result, cases)
+        @result = result
+        @output = nil
+        @matched = false
+
+        cases.each do |name, kase|
+          define_singleton_method(name) do |pattern = nil, &block|
+            handle_case(kase, pattern, &block)
+          end
+        end
+      end
+
+      def call(&block)
+        block.call(self)
+        output
+      end
+
+      private
+
+      def handle_case(kase, pattern, &block)
+        return(output) if matched?
+
+        if kase.matches?(pattern, result)
+          @matched = true
+          @output = block.call(kase.resolve(result))
+        end
```

Comment:
@solnic I hadn't looked into performance yet, no. But I'll look into it.

The reason for the mutation here is so ensure a single match only, and to allow the user's match block to have a proper return value (the return value from the match), even if several non-matches have executed after. I'll look for  ways to do this better :)